codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -284,194 +345,485 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
+                reevaluateWritability();
+                return true;
+            }
+            int numBytesToBeConsumed = 0;
+            if (msg instanceof Http2DataFrame) {
+                numBytesToBeConsumed = ((Http2DataFrame) msg).flowControlledBytes();
+                allocHandle.lastBytesRead(numBytesToBeConsumed);
+            } else {
+                allocHandle.lastBytesRead(ARBITRARY_MESSAGE_SIZE);
+            }
+            allocHandle.incMessagesRead(1);
+            pipeline().fireChannelRead(msg);
+
+            if (numBytesToBeConsumed != 0) {
+                bytesConsumed(numBytesToBeConsumed);
+            }
+            return true;
+        }
+
+        private void reevaluateWritability() {
+            ChannelOutboundBuffer buffer = unsafe().outboundBuffer();
+            // If the buffer is not writable but should be writable, then write and flush a dummy object
+            // to trigger a writability change.
+            if (!buffer.isWritable() && buffer.totalPendingWriteBytes() < config.getWriteBufferHighWaterMark()) {
+                buffer.addMessage(REEVALUATE_WRITABILITY_MESSAGE, 1, voidPromise());
+                unsafe().flush();
+            }
+        }
+
+        private void bytesConsumed(final int bytes) {
+            ctx.write(new DefaultHttp2WindowUpdateFrame(bytes).stream(stream()));
+        }
+
+        private Http2StreamFrame validateStreamFrame(Object msg) {
+            if (!(msg instanceof Http2StreamFrame)) {
+                String msgString = msg.toString();
+                ReferenceCountUtil.release(msg);
+                throw new IllegalArgumentException("Message must be a Http2StreamFrame: " + msgString);
+            }
+            Http2StreamFrame frame = (Http2StreamFrame) msg;
+            if (frame.stream() != null) {
+                String msgString = msg.toString();
+                ReferenceCountUtil.release(frame);
+                throw new IllegalArgumentException("Stream must not be set on the frame: " + msgString);
+            }
+            return frame;
+        }
+
+        private final class Http2ChannelUnsafe extends AbstractUnsafe {
+            @Override
+            public void connect(final SocketAddress remoteAddress,
+                                SocketAddress localAddress, final ChannelPromise promise) {
+                promise.setFailure(new UnsupportedOperationException());
+            }
+        }
+
+        /**
+         * {@link ChannelConfig} so that the high and low writebuffer watermarks can reflect the outbound flow control
+         * window, without having to create a new {@link WriteBufferWaterMark} object whenever the flow control window
+         * changes.
+         */
+        private final class Http2StreamChannelConfig extends DefaultChannelConfig {
+
+            // TODO(buchgr): Overwrite the RecvByteBufAllocator. We only need it to implement max messages per read.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)